{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# file transfer protocol\n",
    "FTP_ROOT = 'ftp://ftp.ncbi.nlm.nih.gov/genomes'\n",
    "ALIGNER_PATH = '/home/smgds/anaconda3/pkgs/bowtie2-2.3.5.1/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# to make graphs\n",
    "from matplotlib import pyplot\n",
    "# work with numeric arrays\n",
    "from numpy import array\n",
    "# run external indexing\n",
    "from subprocess import call\n",
    "\n",
    "\n",
    "# file matching function (searches for extensions)\n",
    "import os\n",
    "# for compressed files\n",
    "import gzip\n",
    "from fnmatch import fnmatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# urlopen communicate with remote data server\n",
    "try:\n",
    "  # Python 3\n",
    "  from urllib.request import urlopen\n",
    "except ImportError:\n",
    "  # Python 2\n",
    "  from urllib2 import urlopen\n",
    "\n",
    "# import HTSeq\n",
    "try:\n",
    "  from HTSeq import FastqReader, SAM_Reader, GFF_Reader\n",
    "  from HTSeq import GenomicArray, GenomicInterval\n",
    "  HAVE_HTSEQ = True\n",
    "except ImportError:\n",
    "  print('HTSeq library inaccessible or not installed. Some examples will not work')\n",
    "  HAVE_HTSEQ = False  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# download function (url of remote file, location of local copy)  \n",
    "def downloadFile(remoteFile, localFile):\n",
    "\n",
    "  print('Downloading to %s ...' % localFile)\n",
    "  \n",
    "  response = urlopen(remoteFile)\n",
    "  data = response.read()\n",
    "  fileObj = open(localFile, 'wb')\n",
    "  fileObj.write(data)\n",
    "  fileObj.close()\n",
    "  \n",
    "  print(' ...done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# searches remote directory for all files with specified format\n",
    "def downloadGenomeFiles(remoteDir, localDir, fileType='*.fna', url=FTP_ROOT):\n",
    "  \n",
    "  if remoteDir[0] != '/':\n",
    "    remoteDir = '/' + remoteDir\n",
    "\n",
    "  if remoteDir[-1] != '/':\n",
    "    remoteDir = remoteDir + '/'\n",
    "  \n",
    "  remotePath = url + remoteDir\n",
    "  print(\"Reading %s\" % remotePath)\n",
    "  \n",
    "  response = urlopen(remotePath)\n",
    "  data = response.read()\n",
    "\n",
    "  fileNames = []\n",
    "  filePaths = []\n",
    "  chromosomeDirs = []\n",
    "  lines = data.split('\\n')\n",
    "\n",
    "  for line in lines:\n",
    "    line = line.strip()\n",
    "    \n",
    "    if not line:\n",
    "      continue\n",
    "\n",
    "    fileName = line.split()[-1]\n",
    "    \n",
    "    if fileName.startswith('CHR'):\n",
    "      chromosomeDirs.append(fileName + '/')\n",
    "\n",
    "    elif fnmatch(fileName, fileType):\n",
    "      fileNames.append(fileName)\n",
    "      continue\n",
    "\n",
    "  for fileName in fileNames:\n",
    "    filePath = os.path.join(localDir, fileName)\n",
    "    filePath = os.path.abspath(filePath)\n",
    "    filePaths.append(filePath)\n",
    "    downloadFile(url + remoteDir + fileName, filePath)\n",
    "\n",
    "  for chromosomeDir in chromosomeDirs:\n",
    "    subDir = remoteDir + chromosomeDir\n",
    "    filePaths += downloadGenomeFiles(subDir, localDir,  fileType, url)\n",
    "\n",
    "  return filePaths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# unzips files ending in .gz  \n",
    "def uncompressGzFile(fileName):\n",
    "\n",
    "  if fileName.endswith('.gz'):\n",
    "    inFileObj = gzip.open(fileName, 'rb')\n",
    " \n",
    "    fileName = fileName[:-3]\n",
    "    outFileObj = open(fileName, 'w')\n",
    "\n",
    "    # writes each uncompressed line into new file\n",
    "    for line in inFileObj:\n",
    "      outFileObj.write(line)\n",
    "    \n",
    "    inFileObj.close()\n",
    "    outFileObj.close()\n",
    "\n",
    "  return fileName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# uses Bowtie to index genome data(binary index file to check short reads against)\n",
    "# input=genome directory, output directory\n",
    "def indexGenome(genomeName, fileNames, outputDir,\n",
    "                tableSize=10, quiet=True, pack=True):\n",
    "\n",
    "  # unzips compressed file\n",
    "  fastaFiles = []\n",
    "  for fileName in fileNames:\n",
    "    fileName = uncompressGzFile(fileName)\n",
    "    fastaFiles.append(fileName)\n",
    "\n",
    "  fastaFileStr= ','.join(fastaFiles)\n",
    "\n",
    "  # executes any command line arguments\n",
    "  cmdArgs = [ALIGNER_PATH+'bowtie-build', '-f']\n",
    "  \n",
    "  if quiet:\n",
    "    cmdArgs.append('-q')\n",
    "  \n",
    "  if pack:\n",
    "    cmdArgs.append('-p')\n",
    "\n",
    "  cmdArgs += ['-t', str(tableSize), fastaFileStr, genomeName]\n",
    "\n",
    "  # prints info about sequence file\n",
    "  print(' '.join(cmdArgs))\n",
    "  call(cmdArgs, cwd=outputDir) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# uses indexed data to align be wrapping Bowtie\n",
    "def genomeAlign(genomeName, genomeDir, readFiles, pairedReadFiles=None, \n",
    "                outFile=None, leftTrim=0, rightTrim=0, useSOLiD=False, \n",
    "                qualType='sanger', maxMismatches=2, highQualLen=28, \n",
    "                pairSepRange=(0,250), showHits=1, maxHits=2,\n",
    "                enforceBest=True, cpuCores=None):\n",
    "\n",
    "  # links variable name to its value\n",
    "  os.environ['BOWTIE_INDEXES'] = genomeDir\n",
    "  \n",
    "  if not outFile:\n",
    "    outFile = genomeName + '.sam'\n",
    "\n",
    "  cmdArgs = [ALIGNER_PATH+'bowtie', '-S']\n",
    "  readFilesStr = ','.join(readFiles)\n",
    "\n",
    "  # identifies format of input file\n",
    "  foreName, fileType = os.path.splitext(readFiles[0])\n",
    "\n",
    "  if fileType in ('fa','fna','mfa','fasta'):\n",
    "    cmdArgs.append('-f')\n",
    "    \n",
    "  elif fileType in ('fq','fastq'):\n",
    "    cmdArgs.append('-q')\n",
    "  \n",
    "  else:\n",
    "    cmdArgs.append('-r')\n",
    "\n",
    "  if useSOLiD:\n",
    "    cmdArgs.append('-C')\n",
    "  \n",
    "  if qualType == 'illumina1.3': # Phred+64\n",
    "    cmdArgs.append('--phred64-quals')\n",
    "  \n",
    "  elif qualType == 'solexa': # Phred+59\n",
    "    cmdArgs.append('--solexa-quals')\n",
    "  \n",
    "  else: # Sanger, current illumina:  Phred + 33\n",
    "    cmdArgs.append('--phred33-quals')  \n",
    "\n",
    "  # reports best match alignments\n",
    "  if enforceBest:\n",
    "    cmdArgs.append('--best')  \n",
    "  \n",
    "  if not cpuCores:\n",
    "    import multiprocessing\n",
    "    cpuCores = multiprocessing.cpu_count()\n",
    "\n",
    "  cmdArgs += ['-5', str(leftTrim),\n",
    "              '-3', str(rightTrim),\n",
    "              '-n', str(maxMismatches),\n",
    "              '-l', str(highQualLen),\n",
    "              '-k', str(showHits),\n",
    "              '-m', str(maxHits),\n",
    "              '-p', str(cpuCores),\n",
    "              '--chunkmbs', '256']\n",
    "\n",
    "  # for second list of file names\n",
    "  if pairedReadFiles:\n",
    "    pairedReadFiles = ','.join(pairedReadFiles)\n",
    "    minSep, maxSep = pairSepRange\n",
    "    cmdArgs += [genomeName,\n",
    "                '--minins', str(minSep),\n",
    "                '--maxins', str(maxSep),\n",
    "                '-1', readFilesStr,\n",
    "                '-2', pairedReadFiles,\n",
    "                outFile]\n",
    "    \n",
    "  else:\n",
    "    cmdArgs += [genomeName, readFilesStr, outFile]\n",
    "\n",
    "  # runs program\n",
    "  call(cmdArgs)  \n",
    "  return outFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    " \n",
    "  # download .fna file\n",
    "  print('\\n File is downloading\\n')\n",
    "\n",
    "  remoteGenome = FTP_ROOT+'/HUMAN_MICROBIOM/Bacteria/Escherichia_coli_MS_200_1_uid47275/NZ_ADUC00000000.scaffold.fna.tgz'\n",
    "  downloadFile(remoteGenome, 'EcoliGenome.fasta')\n",
    "  \n",
    "    # index genome file\n",
    "  print('\\n Indexing downloaded file\\n')\n",
    "\n",
    "  filePath = os.path.abspath('EcoliGenome.fasta')\n",
    "\n",
    "  genomeName = 'E_coli'\n",
    "\n",
    "  indexGenome(genomeName, [filePath,], 'examples')\n",
    "\n",
    "\n",
    "  print('\\n Map sequence reads to E.coli genome\\n')\n",
    "\n",
    "  fastqFile = 'EcoliReads.fastq'\n",
    "\n",
    "  genomeAlign(genomeName, 'examples', [fastqFile], qualType='sanger')\n",
    "\n",
    "\n",
    "  if not HAVE_HTSEQ:\n",
    "    import sys\n",
    "    print('\\n Exiting early. HTSeq library not available.')\n",
    "    sys.exit(0)\n",
    "\n",
    "\n",
    "  print('\\n Using HTSeq library with FASTQ reads\\n')\n",
    "\n",
    "  fileObj = FastqReader(fastqFile)\n",
    "\n",
    "  for seqRead in fileObj:\n",
    "    print(seqRead.name)\n",
    "    print(seqRead.seq)\n",
    "    print(seqRead.get_reverse_complement()[::-1])\n",
    "\n",
    "  numReads = 0.0\n",
    "  meanQual = None\n",
    "\n",
    "  for seqRead in fileObj:\n",
    "    print(seqRead.qual)\n",
    "\n",
    "    if meanQual is None:\n",
    "      meanQual = array(seqRead.qual)\n",
    "    else:\n",
    "      meanQual += seqRead.qual\n",
    "    numReads += 1.0\n",
    "\n",
    "  if numReads:\n",
    "    pyplot.plot(meanQual/numReads)\n",
    "    pyplot.show()\n",
    "\n",
    "\n",
    "  print('\\n Using HTSeq library with SAM alignments\\n')\n",
    "\n",
    "  alignFile = 'EcoliGenomeAlign.sam'\n",
    "  chromosomes = set()\n",
    "\n",
    "  for alignment in SAM_Reader(alignFile):\n",
    " \n",
    "    if alignment.aligned:\n",
    "      seqRead = alignment.read\n",
    "      print(seqRead.name)\n",
    "      print(seqRead.seq)\n",
    "\n",
    "      genomeRegion = alignment.iv\n",
    "      chromo = genomeRegion.chrom\n",
    "      strand = genomeRegion.strand\n",
    "      start = genomeRegion.start\n",
    "      end = genomeRegion.end\n",
    "    \n",
    "      chromosomes.add(chromo)\n",
    "      print(chromo, start, end, strand)\n",
    "\n",
    "\n",
    "  print('\\n HTSeq genomic arrays and intervals\\n')\n",
    "\n",
    "  chromosomes = list(chromosomes)\n",
    "  hitMap = GenomicArray(chromosomes, stranded=True, typecode='i')\n",
    "\n",
    "  for alignment in SAM_Reader(alignFile):\n",
    " \n",
    "    if alignment.aligned:\n",
    "      genomeRegion = alignment.iv\n",
    " \n",
    "      if genomeRegion.strand == '+':\n",
    "        hitMap[genomeRegion] = 1\n",
    "      else:\n",
    "        hitMap[genomeRegion] = -1\n",
    "\n",
    "  chromo = chromosomes[0]\n",
    "  endPoint = 2000000\n",
    "  plusStrand  = GenomicInterval(chromo, 0, endPoint, '+')\n",
    "  minusStrand = GenomicInterval(chromo, 0, endPoint, '-')\n",
    "  bothStrands = GenomicInterval(chromo, 0, endPoint, '.')\n",
    "\n",
    "  pyplot.plot(list(hitMap[plusStrand]))\n",
    "  pyplot.plot(list(hitMap[minusStrand]))\n",
    "  pyplot.show()\n",
    "\n",
    "\n",
    "  print('\\n Using HTSeq to access GFF genome features\\n')\n",
    "\n",
    "  remoteFileName = '/HUMAN_MICROBIOM/Bacteria/Escherichia_coli_MS_200_1_uid47275/NZ_ADUC00000000.scaffold.gff.tgz'\n",
    "  gffFile = 'EcoliGenomeFeatures.gff'\n",
    "  downloadFile(FTP_ROOT+remoteFileName, gffFile)\n",
    "\n",
    "  fileObj = GFF_Reader(gffFile)\n",
    "\n",
    "  for genomeFeature in fileObj:\n",
    "\n",
    "    genomeRegion = genomeFeature.iv\n",
    "\n",
    "    data = (genomeRegion.chrom,\n",
    "            genomeRegion.start,\n",
    "            genomeRegion.end,\n",
    "            genomeRegion.strand)\n",
    "\n",
    "    print('%s %s - %s (%s)' % data)\n",
    "\n",
    "    data = (genomeFeature.name,\n",
    "          genomeFeature.type,\n",
    "          genomeFeature.source)\n",
    " \n",
    "    print('%s %s (%s)' % data)\n",
    "\n",
    "    print(genomeFeature.attr)\n",
    "\n",
    "    \n",
    "  print('\\n Using HTSeq to plot features in genomic arrays\\n')\n",
    "\n",
    "  geneMap = GenomicArray('auto', stranded=False, typecode='O')\n",
    "  genePlot = GenomicArray('auto', stranded=False, typecode='i')\n",
    "\n",
    "  for genomeFeature in fileObj:\n",
    " \n",
    "    if genomeFeature.type == 'gene':\n",
    " \n",
    "      genomeRegion = genomeFeature.iv\n",
    "      geneMap[genomeRegion] = genomeFeature\n",
    "      genePlot[genomeRegion] = 1\n",
    "\n",
    "  for region, feature in geneMap.steps():\n",
    " \n",
    "    if feature:\n",
    " \n",
    "      data = (feature.name,\n",
    "              region.start,\n",
    "              region.end,\n",
    "              feature.iv.strand)\n",
    " \n",
    "      print('%s: %s - %s (%s)' % data)\n",
    "\n",
    "  chromosome = genomeRegion.chrom\n",
    "  region = GenomicInterval(chromosome, 0, 40000, '.')\n",
    "  pyplot.plot(list(genePlot[region]))\n",
    "  pyplot.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
