{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2703,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Amy Noyes\n",
    "# BIOL 51000: Data Systems Life Sciences\n",
    "# Week 5 Coding Assignment: Calculating Root-Mean-Square Deviation (RMSD) and aligning a Structure Ensemble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2704,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import math\n",
    "# import Bio\n",
    "from Bio import PDB\n",
    "from math import pi, sqrt, degrees, acos\n",
    "from matplotlib import pyplot\n",
    "from numpy import array, dot, zeros, ones, cross, sqrt, linalg, exp, identity  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2705,
   "metadata": {},
   "outputs": [],
   "source": [
    "# defines 1 letter amino acid codes\n",
    "amino_acids = {'ALA':'A','CYS':'C','ASP':'D','GLU':'E',\n",
    "                       'PHE':'F','GLY':'G','HIS':'H','ILE':'I',\n",
    "                       'LYS':'K','LEU':'L','MET':'M','ASN':'N',\n",
    "                       'PRO':'P','GLN':'Q','ARG':'R','SER':'S',\n",
    "                       'THR':'T','VAL':'V','TRP':'W','TYR':'Y',\n",
    "                       'G':'G','C':'C','A':'A','T':'T','U':'U'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2706,
   "metadata": {},
   "outputs": [],
   "source": [
    "# atomic numbers commonly found in biological molecules\n",
    "atomic_numbers = {'H':1, 'C':12, 'N':14, 'O':16, 'P':31, 'S':32}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2707,
   "metadata": {},
   "outputs": [],
   "source": [
    "# substitution matrix for protein sequence\n",
    "\n",
    "BLOSUM62 = {'A':{'A': 4,'R':-1,'N':-2,'D':-2,'C': 0,'Q':-1,'E':-1,'G': 0,'H':-2,'I':-1,\n",
    "                 'L':-1,'K':-1,'M':-1,'F':-2,'P':-1,'S': 1,'T': 0,'W':-3,'Y':-2,'V': 0,'X':0},\n",
    "            'R':{'A':-1,'R': 5,'N': 0,'D':-2,'C':-3,'Q': 1,'E': 0,'G':-2,'H': 0,'I':-3,\n",
    "                 'L':-2,'K': 2,'M':-1,'F':-3,'P':-2,'S':-1,'T':-1,'W':-3,'Y':-2,'V':-3,'X':0},\n",
    "            'N':{'A':-2,'R': 0,'N': 6,'D': 1,'C':-3,'Q': 0,'E': 0,'G': 0,'H': 1,'I':-3,\n",
    "                 'L':-3,'K': 0,'M':-2,'F':-3,'P':-2,'S': 1,'T': 0,'W':-4,'Y':-2,'V':-3,'X':0},\n",
    "            'D':{'A':-2,'R':-2,'N': 1,'D': 6,'C':-3,'Q': 0,'E': 2,'G':-1,'H':-1,'I':-3,\n",
    "                 'L':-4,'K':-1,'M':-3,'F':-3,'P':-1,'S': 0,'T':-1,'W':-4,'Y':-3,'V':-3,'X':0},\n",
    "            'C':{'A': 0,'R':-3,'N':-3,'D':-3,'C': 9,'Q':-3,'E':-4,'G':-3,'H':-3,'I':-1,\n",
    "                 'L':-1,'K':-3,'M':-1,'F':-2,'P':-3,'S':-1,'T':-1,'W':-2,'Y':-2,'V':-1,'X':0},\n",
    "            'Q':{'A':-1,'R': 1,'N': 0,'D': 0,'C':-3,'Q': 5,'E': 2,'G':-2,'H': 0,'I':-3,\n",
    "                 'L':-2,'K': 1,'M': 0,'F':-3,'P':-1,'S': 0,'T':-1,'W':-2,'Y':-1,'V':-2,'X':0},\n",
    "            'E':{'A':-1,'R': 0,'N': 0,'D': 2,'C':-4,'Q': 2,'E': 5,'G':-2,'H': 0,'I':-3,\n",
    "                 'L':-3,'K': 1,'M':-2,'F':-3,'P':-1,'S': 0,'T':-1,'W':-3,'Y':-2,'V':-2,'X':0},\n",
    "            'G':{'A': 0,'R':-2,'N': 0,'D':-1,'C':-3,'Q':-2,'E':-2,'G': 6,'H':-2,'I':-4,\n",
    "                 'L':-4,'K':-2,'M':-3,'F':-3,'P':-2,'S': 0,'T':-2,'W':-2,'Y':-3,'V':-3,'X':0},\n",
    "            'H':{'A':-2,'R': 0,'N': 1,'D':-1,'C':-3,'Q': 0,'E': 0,'G':-2,'H': 8,'I':-3,\n",
    "                 'L':-3,'K':-1,'M':-2,'F':-1,'P':-2,'S':-1,'T':-2,'W':-2,'Y': 2,'V':-3,'X':0},\n",
    "            'I':{'A':-1,'R':-3,'N':-3,'D':-3,'C':-1,'Q':-3,'E':-3,'G':-4,'H':-3,'I': 4,\n",
    "                 'L': 2,'K':-3,'M': 1,'F': 0,'P':-3,'S':-2,'T':-1,'W':-3,'Y':-1,'V': 3,'X':0},\n",
    "            'L':{'A':-1,'R':-2,'N':-3,'D':-4,'C':-1,'Q':-2,'E':-3,'G':-4,'H':-3,'I': 2,\n",
    "                 'L': 4,'K':-2,'M': 2,'F': 0,'P':-3,'S':-2,'T':-1,'W':-2,'Y':-1,'V': 1,'X':0},\n",
    "            'K':{'A':-1,'R': 2,'N': 0,'D':-1,'C':-3,'Q': 1,'E': 1,'G':-2,'H':-1,'I':-3,\n",
    "                 'L':-2,'K': 5,'M':-1,'F':-3,'P':-1,'S': 0,'T':-1,'W':-3,'Y':-2,'V':-2,'X':0},\n",
    "            'M':{'A':-1,'R':-1,'N':-2,'D':-3,'C':-1,'Q': 0,'E':-2,'G':-3,'H':-2,'I': 1,\n",
    "                 'L': 2,'K':-1,'M': 5,'F': 0,'P':-2,'S':-1,'T':-1,'W':-1,'Y':-1,'V': 1,'X':0},\n",
    "            'F':{'A':-2,'R':-3,'N':-3,'D':-3,'C':-2,'Q':-3,'E':-3,'G':-3,'H':-1,'I': 0,\n",
    "                 'L': 0,'K':-3,'M': 0,'F': 6,'P':-4,'S':-2,'T':-2,'W': 1,'Y': 3,'V':-1,'X':0},\n",
    "            'P':{'A':-1,'R':-2,'N':-2,'D':-1,'C':-3,'Q':-1,'E':-1,'G':-2,'H':-2,'I':-3,\n",
    "                 'L':-3,'K':-1,'M':-2,'F':-4,'P': 7,'S':-1,'T':-1,'W':-4,'Y':-3,'V':-2,'X':0},\n",
    "            'S':{'A': 1,'R':-1,'N': 1,'D': 0,'C':-1,'Q': 0,'E': 0,'G': 0,'H':-1,'I':-2,\n",
    "                 'L':-2,'K': 0,'M':-1,'F':-2,'P':-1,'S': 4,'T': 1,'W':-3,'Y':-2,'V':-2,'X':0},\n",
    "            'T':{'A': 0,'R':-1,'N': 0,'D':-1,'C':-1,'Q':-1,'E':-1,'G':-2,'H':-2,'I':-1,\n",
    "                 'L':-1,'K':-1,'M':-1,'F':-2,'P':-1,'S': 1,'T': 5,'W':-2,'Y':-2,'V': 0,'X':0},\n",
    "            'W':{'A':-3,'R':-3,'N':-4,'D':-4,'C':-2,'Q':-2,'E':-3,'G':-2,'H':-2,'I':-3,\n",
    "                 'L':-2,'K':-3,'M':-1,'F': 1,'P':-4,'S':-3,'T':-2,'W':11,'Y': 2,'V':-3,'X':0},\n",
    "            'Y':{'A':-2,'R':-2,'N':-2,'D':-3,'C':-2,'Q':-1,'E':-2,'G':-3,'H': 2,'I':-1,\n",
    "                 'L':-1,'K':-2,'M':-1,'F': 3,'P':-3,'S':-2,'T':-2,'W': 2,'Y': 7,'V':-1,'X':0},\n",
    "            'V':{'A': 0,'R':-3,'N':-3,'D':-3,'C':-1,'Q':-2,'E':-2,'G':-3,'H':-3,'I': 3,\n",
    "                 'L': 1,'K':-2,'M': 1,'F':-1,'P':-2,'S':-2,'T': 0,'W':-3,'Y':-1,'V': 4,'X':0},\n",
    "            'X':{'A': 0,'R': 0,'N': 0,'D': 0,'C': 0,'Q': 0,'E': 0,'G': 0,'H': 0,'I': 0,\n",
    "                 'L': 0,'K': 0,'M': 0,'F': 0,'P': 0,'S': 0,'T': 0,'W': 0,'Y': 0,'V': 0,'X':0}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2708,
   "metadata": {},
   "outputs": [],
   "source": [
    "# substitution matrix for DNA\n",
    "DNA_1 = {'G': { 'G': 1, 'C':-3, 'A':-3, 'T':-3, 'N':0 },\n",
    "         'C': { 'G':-3, 'C': 1, 'A':-3, 'T':-3, 'N':0 },\n",
    "         'A': { 'G':-3, 'C':-3, 'A': 1, 'T':-3, 'N':0 },\n",
    "         'T': { 'G':-3, 'C':-3, 'A':-3, 'T': 1, 'N':0 },\n",
    "         'N': { 'G': 0, 'C': 0, 'A': 0, 'T': 0, 'N':0 }}  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2709,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import Protein Data Bank files directly from PDB\n",
    "# import urlopen function\n",
    "try:\n",
    "  # for python 3\n",
    "  from urllib.request import urlopen\n",
    "except ImportError:\n",
    "  # for python 2\n",
    "  from urllib2 import urlopen\n",
    "    \n",
    "# update PDB url as needed\n",
    "PDB_URL = 'http://files.rcsb.org/view/' \\\n",
    "          '%s.pdb?format=PDB&compression=None'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2710,
   "metadata": {},
   "outputs": [],
   "source": [
    "# download PDB file by ID and filename if available\n",
    "def download_PDB(pdbId, fileName=None):\n",
    "\n",
    "  # for no filename: create filename by appending .pdb to ID\n",
    "  if not fileName:\n",
    "    fileName = '%s.pdb' % pdbId\n",
    "\n",
    "  # retrieve file\n",
    "  retrieved = urlopen(PDB_URL % pdbId)\n",
    "  # convert bytes to string using UTF-8  \n",
    "  data = retrieved.read().decode('utf-8')\n",
    "\n",
    "  # write download to file\n",
    "  fileObj = open(fileName, 'w')\n",
    "  fileObj.write(data)\n",
    "  fileObj.close()\n",
    "\n",
    "  return fileName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2711,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creates structure object from PDB file\n",
    "def get_structures(fileName):\n",
    "\n",
    "  fileObject = open(fileName)\n",
    "  structure = None\n",
    "  name = 'unknown'\n",
    "  conformation = 0\n",
    "  pdbId = None\n",
    "  structures = []\n",
    "\n",
    "  # loops through each line\n",
    "  for line in fileObject:\n",
    "    record = line[0:6].strip()\n",
    "    \n",
    "    if record == 'HEADER':\n",
    "      pdbId = line.split()[-1]\n",
    "\n",
    "    elif record == 'TITLE':\n",
    "      name = line[10:].strip()\n",
    "\n",
    "    elif record == 'MODEL':\n",
    "      conformation = int(line[10:14])\n",
    "\n",
    "    elif record == 'ENDMDL':\n",
    "      structure = None\n",
    "\n",
    "    elif record == 'ATOM':\n",
    "      atomName   = line[12:16].strip()\n",
    "      resName    = line[17:20].strip()\n",
    "      chain_code  = line[21:22].strip()\n",
    "      seqId      = int(line[22:26])\n",
    "      x          = float(line[30:38])\n",
    "      y          = float(line[38:46])\n",
    "      z          = float(line[46:54])\n",
    "      segment    = line[72:76].strip()\n",
    "      element    = line[76:78].strip() \n",
    "      \n",
    "      # create new chain  \n",
    "      if chain_code == '':\n",
    "        if segment:\n",
    "          chain_code = segment\n",
    "        else:\n",
    "          chain_code = 'A'\n",
    "\n",
    "      if not structure:\n",
    "        structure = Structure(name, conformation, pdbId)\n",
    "        structures.append(structure)\n",
    "\n",
    "      chain = structure.get_chain(chain_code)\n",
    "      if not chain:\n",
    "        chain = Chain(structure, chain_code)\n",
    "\n",
    "      # create new residue  \n",
    "      residue = chain.get_residue(seqId)\n",
    "      if not residue:\n",
    "        residue = Residue(chain, seqId, resName)\n",
    "      \n",
    "     # missing element -> guess using first character of the atom name\n",
    "      if not element:\n",
    "        element = name[0]\n",
    "\n",
    "      coords = (x,y,z)\n",
    "      atom = Atom(residue, atomName, coords, element)\n",
    "    \n",
    "  fileObject.close()\n",
    "\n",
    "  return structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2712,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3D group of molecules\n",
    "class Structure:\n",
    "    \n",
    "  # creates instance of self  \n",
    "  def __init__(self, name, conformation=0, pdbId=None):\n",
    "    # exception for blank name\n",
    "    if not name:\n",
    "      raise Exception('Name is blank')\n",
    "    self._name = name\n",
    "    self.conformation = conformation\n",
    "    self.pdbId = pdbId\n",
    "    # links structure to children (chains)\n",
    "    self.chains = []\n",
    "\n",
    "  # chains delete when structure disappears\n",
    "  def delete(self):\n",
    "    for chain in self.chains:\n",
    "      chain.delete()\n",
    "    \n",
    "  # find chain with matching atribute (loops through list of children)\n",
    "  def get_chain(self, code):\n",
    "    for chain in self.chains:\n",
    "      if chain.code == code:\n",
    "        return chain\n",
    "\n",
    "    return None\n",
    "\n",
    "  # gets a list of pdbIds\n",
    "  def get_pdb_ids(self):\n",
    "    return list(self.pdbIds)\n",
    "\n",
    "  # gets molecular mass of structure\n",
    "  def get_mass(self):\n",
    "    if hasattr(self, 'mass'):\n",
    "      return self.mass\n",
    "\n",
    "    mass = 0\n",
    "    for chain in self.chains:\n",
    "      mass += chain.get_mass()\n",
    "    self.mass = mass # cache for next time\n",
    "    return mass\n",
    "\n",
    "  # name of molecule as string\n",
    "  def get_name(self):\n",
    "    return self._name\n",
    "\n",
    "  def set_name(self, name):\n",
    "    if not name:\n",
    "      raise Exception('Name is blank')\n",
    "    self._name = name\n",
    "\n",
    "  name = property(get_name, set_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2713,
   "metadata": {},
   "outputs": [],
   "source": [
    "# child of Structure\n",
    "# each molecule defined by chain class\n",
    "\n",
    "class Chain:\n",
    "  allowed_mol_types = ('protein', 'DNA', 'RNA')\n",
    "\n",
    "  \n",
    "  def __init__(self, structure, code, mol_type='protein'):\n",
    "    if not code:\n",
    "      raise Exception('Error. Code is blank.')\n",
    "    \n",
    "    if mol_type not in self.allowed_mol_types:\n",
    "      raise Exception('Incorrect molecule type' %\n",
    "                      (mol_type, self.allowed_mol_types))\n",
    "\n",
    "    # checks for unique code\n",
    "    chain = structure.get_chain(code)\n",
    "    if chain:\n",
    "      raise Exception('Code already used.' % code)\n",
    "\n",
    "    # add chain to structure.chains list\n",
    "    self.structure = structure\n",
    "    self.code = code\n",
    "    self._mol_type = mol_type\n",
    "    self.resDict = {}\n",
    "    self.residues = []\n",
    "    structure.chains.append(self)\n",
    "\n",
    "  # removes from structure list of children\n",
    "  def delete(self):\n",
    "    for residue in self.residues:\n",
    "      residue.delete()\n",
    "    self.structure.chains.remove(self)\n",
    "\n",
    "  # returns child residue with given seqId\n",
    "  def get_residue(self, seqId):\n",
    "    return self.resDict.get(seqId)\n",
    "\n",
    "  # returns atoms of all residues\n",
    "  def get_atoms(self):\n",
    "    atoms = []\n",
    "    for residue in self.residues:\n",
    "      atoms.extend(residue.atoms)\n",
    "    return atoms\n",
    "    \n",
    "  # returns molecule type  \n",
    "  def get_mol_type(self):\n",
    "    return self._mol_type\n",
    "\n",
    "  mol_type = property(get_mol_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2714,
   "metadata": {},
   "outputs": [],
   "source": [
    "# child of Chain. seqId links to Chain\n",
    "class Residue:\n",
    "    \n",
    "  def __init__(self, chain, seqId, code=None):\n",
    "    if not seqId:\n",
    "      raise Exception('seqId is blank')\n",
    "    residue = chain.get_residue(seqId)\n",
    "    if residue:\n",
    "      raise Exception('seqId=\"%s\" already used' % seqId)\n",
    "\n",
    "    # add residues to chain.residues list\n",
    "    self.chain = chain\n",
    "    self.seqId = seqId\n",
    "    self.code = code\n",
    "    self.atomDict = {}\n",
    "    self.atoms = []\n",
    "    chain.resDict[seqId] = self\n",
    "    chain.residues.append(self)\n",
    "\n",
    "  # removes from chain\n",
    "  def delete(self):\n",
    "    for atom in self.atoms:\n",
    "      atom.delete()\n",
    "    del self.chain.resDict[self.seqId]\n",
    "    self.chain.residues.remove(self)\n",
    "\n",
    "  def get_atom(self, name):\n",
    "    return self.atomDict.get(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2715,
   "metadata": {},
   "outputs": [],
   "source": [
    "# child of Residue. \n",
    "class Atom:\n",
    "  def __init__(self, residue, name, coords, element):\n",
    "    if not name:\n",
    "      raise Exception('Name is blank')\n",
    "    atom = residue.get_atom(name)\n",
    "    if atom:\n",
    "      raise Exception('name=\"%s\" already used' % name)\n",
    "    if len(coords) != 3:\n",
    "      raise Exception('Coordinates must contain three values')\n",
    " \n",
    "    self.residue = residue\n",
    "    self.name = name\n",
    "    self.coords = array(coords)\n",
    "    self.element = element\n",
    "\n",
    "    residue.atomDict[name] = self\n",
    "    residue.atoms.append(self)\n",
    "\n",
    "  def delete(self):\n",
    "    del self.residue.atomDict[self.name]\n",
    "    self.residue.atoms.remove(self)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2716,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculates center of mass, zero point\n",
    "def center_coords(coords, weights):\n",
    "\n",
    "  wCoords = coords.transpose() * weights\n",
    "\n",
    "  # sum xyz total weights\n",
    "  xyzTotals = wCoords.sum(axis=1)\n",
    "\n",
    "  # average weight -> center\n",
    "  center = xyzTotals/sum(weights)\n",
    "\n",
    "  # remove previous center\n",
    "  coords -= center\n",
    "  \n",
    "  return coords, center  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2717,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculates best rotation superimposing two arrays\n",
    "def align_coords(coordsA, coordsB, weights=None):\n",
    "\n",
    "  # no weights, new weights created with series of one's\n",
    "  n = len(coordsA)\n",
    "  if weights is None:\n",
    "    weights = ones(n)\n",
    "\n",
    "  # initial matrix -> weighted dot product of 1 coordinate array \n",
    "  # and transpose of 2nd array\n",
    "  rMat = dot(coordsB.transpose()*weights, coordsA)\n",
    "\n",
    "  # singular value decomposition -> rotatio matrix, \n",
    "  # linear scaling factors, opposing rotation matrix\n",
    "  rMat1, scales, rMat2 = linalg.svd(rMat)\n",
    "\n",
    "  # checks for mirror image\n",
    "  sign = linalg.det(rMat1) * linalg.det(rMat2)\n",
    "\n",
    "  if sign < 0:\n",
    "    rMat1[:,2] *= -1\n",
    "  \n",
    "  # optimized rotation matrix from 2 matrices from linalg.svd\n",
    "  rotation = dot(rMat1, rMat2)\n",
    "    \n",
    "  # applies the transformation  \n",
    "  coordsB = dot(coordsB, rotation)\n",
    "  \n",
    "  return rotation, coordsB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2718,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculates root mean square deviation\n",
    "# input reference coordinates, list of arrays to compare, list of weights\n",
    "\n",
    "def calc_rmsds(ref_coord, all_coor, weights):\n",
    "\n",
    "  rmsds = []\n",
    "  # sum all weights\n",
    "  totalWeight = sum(weights)\n",
    "  # array of zeros same size as coordinate arrays\n",
    "  totalSquares = zeros(ref_coord.shape)\n",
    "  \n",
    "  # loop through list of arrays and compare to reference\n",
    "  for coords in all_coor:\n",
    "    delta = coords-ref_coord\n",
    "    squares = delta * delta\n",
    "    totalSquares += squares\n",
    "    # sum of the square values along the spatial axis\n",
    "    sumSquares = weights*squares.sum(axis=1)\n",
    "    # average atomic square deviation\n",
    "    rmsds.append( sqrt(sum(sumSquares)/totalWeight) )\n",
    "  \n",
    "  # each atomâ€™s rmsd over whole set of conformations\n",
    "  nStruct = len(all_coor)\n",
    "  atom_rmsds = sqrt(totalSquares.sum(axis=1)/nStruct)\n",
    "  \n",
    "  return rmsds, atom_rmsds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2719,
   "metadata": {},
   "outputs": [],
   "source": [
    "# superimposes all conformations of a structure ensemble\n",
    "def superimpose_coords(all_coor, weights, threshold=5.0):\n",
    "\n",
    "  nStruct = len(all_coor)\n",
    "\n",
    "  # reference set to first array\n",
    "  ref_coord = all_coor[0]\n",
    "  # create array of zeros of same size\n",
    "  mean_coord = zeros(ref_coord.shape)\n",
    "  rotations = []\n",
    "\n",
    "  # loop through all arrays and align to reference\n",
    "  for index, coords in enumerate(all_coor):\n",
    "    # if index is zero an identity matrix is created\n",
    "    if index == 0:\n",
    "      rotation = identity(3)\n",
    "    else:  \n",
    "      rotation, coords = align_coords(ref_coord, coords, weights)\n",
    "      all_coor[index] = coords # Update to aligned\n",
    "      \n",
    "    # rotation matrix is added to rotations list\n",
    "    rotations.append(rotation)\n",
    "    # coordinate array added to mean_coords\n",
    "    mean_coord += coords\n",
    "  \n",
    "  # average positions\n",
    "  mean_coord /= nStruct\n",
    "  \n",
    "  # calculate RMSD values\n",
    "  rmsds, atom_rmsds = calc_rmsds(mean_coord, all_coor, weights)  \n",
    "  \n",
    "  # finds smallest rmds value\n",
    "  best_rmsd = min(rmsds)\n",
    "  best_index = rmsds.index(best_rmsd)\n",
    "  best_coords = all_coor[best_index]\n",
    "\n",
    "  # rmds value is scaled by a threshold to assign sensitivity to structural variation\n",
    "  weight_scale = atom_rmsds/threshold\n",
    "  # atoms with small rmsd have largest weights\n",
    "  # as variance increases the weight diminishes exponentially\n",
    "  weights = (weights * (exp(-weight_scale*weight_scale)))\n",
    "  \n",
    "  # copies coordinate array for average coordinates\n",
    "  mean_coord = best_coords.copy()\n",
    "  \n",
    "  # loop of all coordinates to compare to average\n",
    "  for index, coords in enumerate(all_coor):\n",
    "    if index != best_index:\n",
    "      rotation, coords = align_coords(best_coords, coords, weights)\n",
    "      rotations[index] = rotation\n",
    "      all_coor[index] = coords\n",
    "      mean_coord += coords\n",
    " \n",
    "  # calculate coordinate average\n",
    "  mean_coord /= nStruct \n",
    "\n",
    "  # calculates rmsd of structures and individual atoms\n",
    "  weights = ones(len(weights))\n",
    "  rmsds, atom_rmsds = calc_rmsds(mean_coord, all_coor, weights)\n",
    "  \n",
    "  return all_coor, rmsds, atom_rmsds, rotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2720,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extracts coordinate arrays, superimposes, updates original array\n",
    "def super_impose(structures):\n",
    "\n",
    "  weights = None\n",
    "  all_coor = []\n",
    "\n",
    "  # convert atom coordinates and atom list to NumPy array\n",
    "  for structure in structures:\n",
    "    atoms, coords = get_atom_coords(structure)\n",
    "    \n",
    "    if weights is None:\n",
    "      weights = [atomic_numbers[atom.element] for atom in atoms]\n",
    "      weights = array(weights)\n",
    "\n",
    "    # center coordinates according to weight\n",
    "    coords, center = center_coords(coords, weights)\n",
    "    all_coor.append(coords)\n",
    "  \n",
    "  # superimpose arrays\n",
    "  results = superimpose_coords(all_coor, weights)\n",
    "  # output: rmsd for atom and structure and list of rotaations\n",
    "  all_coor, rmsds, atom_rmsds, rotations = results  \n",
    "  \n",
    "  # update atom objects with new coordinates\n",
    "  for i, structure in enumerate(structures):\n",
    "    atoms, oldCoords = get_atom_coords(structure)\n",
    "    \n",
    "    for j, atom in enumerate(atoms):\n",
    "      atom.coords = all_coor[i][j]\n",
    "\n",
    "  return rmsds, atom_rmsds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2721,
   "metadata": {},
   "outputs": [],
   "source": [
    "# converts 3 letter amino acids to 1 letter\n",
    "\n",
    "def get_chain_sequence(chain):\n",
    "\n",
    "  letters = []\n",
    "  \n",
    "  # loops through residues in input chain\n",
    "  for residue in chain.residues:\n",
    "    code =  residue.code\n",
    "    # X for unknown amino acid (not in amino_acids dictionary)\n",
    "    letter = amino_acids.get(code, 'X')\n",
    "    letters.append(letter)\n",
    "    \n",
    "  seq = ''.join(letters)\n",
    "\n",
    "  return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2722,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pairwise sequence alignment\n",
    "# imput: seqA and seqB; substitution matrix, gap penalties\n",
    "\n",
    "def sequence_align(seqA, seqB, simMatrix=DNA_1, insert=8, extend=4):\n",
    "\n",
    "  # define comparison grid: +1 larger then length of sequences\n",
    "  numI = len(seqA) + 1\n",
    "  numJ = len(seqB) + 1\n",
    "\n",
    "  # create route matrix\n",
    "  scoreMatrix = [[0] * numJ for x in range(numI)]\n",
    "  routeMatrix = [[0] * numJ for x in range(numI)]\n",
    "  \n",
    "  # edges of matrix at zero\n",
    "  for i in range(1, numI):\n",
    "    routeMatrix[i][0] = 1\n",
    "  \n",
    "  for j in range(1, numJ):\n",
    "    routeMatrix[0][j] = 2\n",
    "  \n",
    "  # fill in remaining matrix\n",
    "  for i in range(1, numI):\n",
    "    for j in range(1, numJ):\n",
    "    \n",
    "      # insertion penalty unless previous has gap then extension penalty\n",
    "      penalty1 = insert\n",
    "      penalty2 = insert\n",
    "      \n",
    "      if routeMatrix[i-1][j] == 1:\n",
    "        penalty1 = extend\n",
    "        \n",
    "      elif routeMatrix[i][j-1] == 2:\n",
    "        penalty2 = extend\n",
    "\n",
    "      # similarity score\n",
    "      similarity = simMatrix[ seqA[i-1] ][ seqB[j-1] ]\n",
    "        \n",
    "      # list of 3 possible paths\n",
    "      paths = [scoreMatrix[i-1][j-1] + similarity, # Route 0\n",
    "               scoreMatrix[i-1][j] - penalty1, # Route 1\n",
    "               scoreMatrix[i][j-1] - penalty2] # Route 2                     \n",
    "      \n",
    "      # maximum value is best score\n",
    "      best = max(paths)\n",
    "      route = paths.index(best)           \n",
    "\n",
    "      scoreMatrix[i][j] = best\n",
    "      routeMatrix[i][j] = route\n",
    "\n",
    "  # add in penalties\n",
    "  alignA = []\n",
    "  alignB = []\n",
    "  \n",
    "  i = numI-1\n",
    "  j = numJ-1\n",
    "  score = scoreMatrix[i][j]\n",
    "\n",
    "    \n",
    "  while i > 0 or j > 0:\n",
    "    route = routeMatrix[i][j]    \n",
    "\n",
    "    # best score from comparing 2 residues\n",
    "    if route == 0: \n",
    "      alignA.append( seqA[i-1] )\n",
    "      alignB.append( seqB[j-1] )\n",
    "      i -= 1\n",
    "      j -= 1\n",
    "\n",
    "    # best score from inserting a gap\n",
    "    elif route == 1: # Gap in seqB\n",
    "      alignA.append( seqA[i-1] )\n",
    "      alignB.append( '-' )\n",
    "      i -= 1 \n",
    "    \n",
    "    # best score from inserting a gap\n",
    "    elif route == 2: # Gap in seqA\n",
    "      alignA.append( '-' )\n",
    "      alignB.append( seqB[j-1] ) \n",
    "      j -= 1\n",
    "  \n",
    "  # correct order of lists\n",
    "  alignA.reverse()\n",
    "  alignB.reverse()\n",
    "  alignA = ''.join(alignA)\n",
    "  alignB = ''.join(alignB)\n",
    "\n",
    "  return score, alignA, alignB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2723,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function for sequence alignment of corresponding residues\n",
    "def backbone_align(chainA, chainB, \n",
    "                              atomNames=set(['CA','C','N']),\n",
    "                              simMatrix=BLOSUM62):\n",
    "\n",
    "  # input chains and list of residue objects\n",
    "  structureA = chainA.structure\n",
    "  structureB = chainB.structure\n",
    "  residuesA = chainA.residues\n",
    "  residuesB = chainB.residues\n",
    "\n",
    "  # get 1 letter sequence strings from Chain\n",
    "  seqA = get_chain_sequence(chainA)\n",
    "  seqB = get_chain_sequence(chainB)\n",
    "\n",
    "  score, alignA, alignB = sequence_align(seqA, seqB, simMatrix)\n",
    "\n",
    "  # list to hold aligned residues\n",
    "  pairedPosA = []\n",
    "  pairedPosB = []  \n",
    "  posA = 0\n",
    "  posB = 0\n",
    "    \n",
    "\n",
    "  # loop to find residues in alignment string from sequence_align\n",
    "  # gaps increase counter +1\n",
    "  for i in range(len(alignA)):\n",
    " \n",
    "    if alignA[i] == '-':\n",
    "      posB += 1\n",
    "   \n",
    "    elif alignB[i] == '-':\n",
    "      posA += 1\n",
    "    \n",
    "    else:\n",
    "      pairedPosA.append(posA)\n",
    "      pairedPosB.append(posB)\n",
    "      \n",
    "      posA += 1\n",
    "      posB += 1\n",
    "\n",
    "  # uses positions to place seqId in list\n",
    "  filterIdsA = [residuesA[p].seqId for p in pairedPosA]\n",
    "  filterIdsB = [residuesB[p].seqId for p in pairedPosB]\n",
    "\n",
    "\n",
    "  # filter to contain backbone structures\n",
    "  backboneStrucA = filter_structure(structureA, [chainA.code], \n",
    "                                      filterIdsA, atomNames)\n",
    "  backboneStrucB = filter_structure(structureB, [chainB.code], \n",
    "                                      filterIdsB, atomNames)\n",
    "  \n",
    "\n",
    "  # list of atoms and coordinate array\n",
    "  atomsA, coordsA = get_atom_coords(backboneStrucA)\n",
    "  atomsB, coordsB = get_atom_coords(backboneStrucB)\n",
    "  weights = ones(len(atomsA))\n",
    "\n",
    "\n",
    "  # coordinates are centered\n",
    "  coordsA, centerA = center_coords(coordsA, weights)\n",
    "  coordsB, centerB = center_coords(coordsB, weights)\n",
    "\n",
    "  # superiposes coordinates\n",
    "  coords = [coordsA, coordsB]\n",
    "  c, rmsds, atom_rmsds, rotations = superimpose_coords(coords, weights)\n",
    "\n",
    "  # rotations and center transforms original structures\n",
    "  rotated_structure(structureA, rotations[0], -centerA)  \n",
    "  rotated_structure(structureB, rotations[1], -centerB)  \n",
    "    \n",
    "  return rmsds, atom_rmsds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2724,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generates rotation matrix\n",
    "# uses axis direction and angle\n",
    "\n",
    "def get_rotation_matrix(axis, angle):\n",
    "  vLen = math.sqrt(sum([xyz*xyz for xyz in axis]))\n",
    "  x, y, z = [xyz/vLen for xyz in axis]\n",
    "  c = math.cos(angle)\n",
    "  d = 1-c\n",
    "  s = math.sin(angle)\n",
    "\n",
    "  R = [[c+d*x*x,   d*x*y-s*z, d*x*z+s*y],\n",
    "       [d*y*x+s*z, c+d*y*y,   d*y*z-s*x],\n",
    "       [d*z*x-s*y, d*z*y+s*x, c+d*z*z  ]]\n",
    "    \n",
    "  return R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2725,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rotates structure while keeping relative position of coordinates\n",
    "def rotate_structure_NumPy(structure, axis=array([2,1,1]), angle=0):\n",
    "\n",
    "  # convert to array\n",
    "  rotated_matrix = array(get_rotation_matrix(axis, angle))\n",
    "\n",
    "  # get atom array with coordinates\n",
    "  atoms, coords = get_atom_coords(structure)\n",
    "  \n",
    "  coords = dot(coords, rotated_matrix.T)\n",
    "  \n",
    "  for index, atom in enumerate(atoms):\n",
    "    atom.coords = list(coords[index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2726,
   "metadata": {},
   "outputs": [],
   "source": [
    "# transforms coordinates\n",
    "# uses identity matrix with size 3 (to match coordinates)\n",
    "def rotated_structure(structure, transform=identity(3),\n",
    "                             translate=(0,0,0)):\n",
    "\n",
    "  # get atom array with coordinates\n",
    "  atoms, coords = get_atom_coords(structure)\n",
    "\n",
    "  # apply transformation\n",
    "  coords = dot(coords, transform.T)\n",
    "  \n",
    "  # apply translation\n",
    "  coords = coords + translate\n",
    "\n",
    "  for index, atom in enumerate(atoms):\n",
    "    atom.coords = coords[index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2727,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes only required atoms\n",
    "def filter_structure(structure, chain_codes=None,\n",
    "                       residueIds=None, atomNames=None):\n",
    "\n",
    "  name = structure.name + '_filter'\n",
    "\n",
    "  # checks if chain, residue, and atom names are defined\n",
    "  if chain_codes:\n",
    "    name += ' ' + ','.join(chain_codes)\n",
    "    chain_codes = set(chain_codes)\n",
    "  \n",
    "  if residueIds:\n",
    "    name += ' ' + ','.join([str(x) for x in residueIds])\n",
    "    residueIds = set(residueIds)\n",
    "\n",
    "  if atomNames:\n",
    "    name += ' ' + ','.join(atomNames)\n",
    "    atomNames = set(atomNames)\n",
    "\n",
    "    \n",
    "  # create copy of structure\n",
    "  conf = structure.conformation\n",
    "  pdbId = structure.pdbId\n",
    "  filterStruc = Structure(name=name, conformation=conf, pdbId=pdbId)\n",
    "\n",
    "    \n",
    "  # loop for selecting chains, residues, and atoms\n",
    "  for chain in structure.chains:\n",
    "    if chain_codes and (chain.code not in chain_codes):\n",
    "      continue\n",
    "  \n",
    "    includeResidues = []\n",
    "    \n",
    "    for residue in chain.residues:\n",
    "      if residueIds and (residue.seqId not in residueIds):\n",
    "        continue\n",
    "\n",
    "      includeAtoms = []\n",
    "\n",
    "      for atom in residue.atoms:\n",
    "        if atomNames and (atom.name not in atomNames):\n",
    "          continue\n",
    "\n",
    "        includeAtoms.append(atom)\n",
    "\n",
    "      if includeAtoms:\n",
    "        includeResidues.append( (residue, includeAtoms) )\n",
    "\n",
    "        \n",
    "    # creates new chain\n",
    "    if includeResidues:\n",
    "        \n",
    "      filterChain = Chain(filterStruc, chain.code, chain.mol_type)\n",
    "      \n",
    "      for residue, atoms in includeResidues:\n",
    "        filterResidue = Residue(filterChain, residue.seqId,\n",
    "                                residue.code)\n",
    "    \n",
    "        for atom in atoms:\n",
    "          coords = array(atom.coords)\n",
    "          Atom(filterResidue, atom.name, coords, atom.element)\n",
    "  \n",
    "\n",
    "  return filterStruc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2728,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_atom_coords(structure):\n",
    "\n",
    "  coords = []\n",
    "  atoms = []\n",
    "  \n",
    "  for chain in structure.chains:\n",
    "    for residue in chain.residues:\n",
    "      for atom in residue.atoms:\n",
    "        coords.append(atom.coords)\n",
    "        atoms.append(atom)\n",
    "\n",
    "  return atoms, array(coords) \n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2729,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Test Root-Mean_Square-Deviation \n",
      "Structure A and Rotated Structure A:\n",
      "[3.5437596409757814e-14, 3.5435795315108795e-14]\n",
      "\n",
      "Superimpose Two Homologous Structures\n",
      "[0.5572256141745706, 0.5572256141745703]\n"
     ]
    }
   ],
   "source": [
    "if __name__ == '__main__':\n",
    "\n",
    "  print('\\nTest Root-Mean_Square-Deviation \\nStructure A and Rotated Structure A:')\n",
    " \n",
    "  strucA = get_structures(download_PDB('5UDO'))[0]\n",
    "  strucB = get_structures(download_PDB('5UDO'))[0]\n",
    "  # rotate file to get strucB\n",
    "  rotate_structure_NumPy(strucA, (1,0,0), pi/3)\n",
    "  rmsds, atom_rmsds = super_impose([strucA, strucB])\n",
    "  print(rmsds) \n",
    " \n",
    "\n",
    "  print('\\nSuperimpose Two Homologous Structures')\n",
    "  # input two homologous structures\n",
    "  struc1 = get_structures(download_PDB('5UDO'))[0]\n",
    "  struc2 = get_structures(download_PDB('5U96'))[0]\n",
    "  chain1 = struc1.get_chain('A')\n",
    "  chain2 = struc2.get_chain('A')\n",
    "  rmsds, atom_rmsds = backbone_align(chain1, chain2)\n",
    "\n",
    "  print(rmsds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
